### DISCUSSION TITLE
Dispose after removal, with .DisposeMany()

### DISCUSSION THREAD

**Discussion by JakenVeina** [reactivemarbles/DynamicData#732] [Ideas]
During debugging of a disposal issue that turned out to be related to something else, I noticed that `.DisposeMany()` does not behave the way I thought it would: items being removed from collections get disposed by the operator before they are removed from downstream collections, instead of after, making it potentially possible for already-disposed items to be invoked in some way. This is due to the use of `.Do()` within `.DisposeMany()`.

https://github.com/reactivemarbles/DynamicData/blob/b3180bc891c49a880d224cabe73f9524e5b8f8c3/src/DynamicData/Cache/Internal/DisposeMany.cs#L34

`.Do()` effectively amounts to...

```cs
public static IObservable<T> Do<T>(this IObservable<T> source, Action<T> onNext)
    => Observable.Create<T>(observer => source.Subscribe(value => 
    {
        onNext.Invoke(value);
        observer.OnNext(value);
    });
```
    
Instead, for `.DisposeMany()` it could be...

```cs
public static IObservable<T> Do<T>(this IObservable<T> source, Action<T> onNext)
    => Observable.Create<T>(observer => source.Subscribe(value => 
    {
        observer.OnNext(value);
        onNext.Invoke(value);
    });
```

This would ensure that disposal doesn't occur until after all downstream collections have processed the removals (at least assuming there are no downstream scheduler changes, which would potentially break `.DisposeMany()` no matter what).

**Comment by dwcullop** [comment]
You have the problem and the solution .... Now all you need is a unit test or two and you'll have yourself a PR!

**Reply by JakenVeina** [reply]
Indeed.

Is it considered normal to just submit un-asked-for PRs for enhancements to a project you've never contributed to before, without any input from maintainers? I feel like I'd find it mildly rude, if it were my project.

**Reply by dwcullop** [reply]
I don't know.  I haven't contributed to that many projects, and I had a similar feeling as you about contributing to this one.  However, so far, Roland has been very receptive.

If you really prefer to be formal about it, then do something about the "input from maintainers" part.  Create an issue to document what you've found.  That will begin a dialog in which you can mention what you did to fix it and ask if a PR would be welcome.

**Reply by JakenVeina** [reply]
> Create an issue to document what you've found. That will begin a dialog in which you can mention what you did to fix it and ask if a PR would be welcome.

That would be what this discussion was for.

**Reply by RolandPheasant** [reply]
A PR would be welcome, and it would be sufficient to reference this issue.  

Also worth considering this too: https://github.com/reactivemarbles/DynamicData/issues/668

**Reply by dwcullop** [reply]
Instead of a special version of `Do`, you could just handle it in the Subscribe lambda.  This issue and #668 could be fixed if Line 34 was changed to something like:

```csharp
    var subscriber = _source.Synchronize(locker).Subscribe(changes =>
                                                           {
                                                               observer.OnNext(changes);
                                                               RegisterForRemoval(changes, cache);
                                                           },
                                                           observer.OnError,
                                                           observer.OnCompleted);
```

**Reply by JakenVeina** [reply]
For sure, I wouldn't write a new `.Do` extension just for this, I'd implement it directly, either like that, or with `Observable.Create()`.

@RolandPheasant do you think a new version of `.DisposeMany()` is warranted, or would a breaking change to the existing method be acceptable?

**Reply by RolandPheasant** [reply]
@JakenVeina I think your observation that objects should be disposed after they have been  removed it correct.  It makes sense.  I remember years ago getting a disposed object access exception but in that case the screen had a long delay on the UI before the object would be removed ie before being removed the row would be highlighted for a couple seconds then removed. My solution at the time was to do the dispose on a delayed stream, but that was probably a different use case.

The suggestion by @dwcullop  would be an easy and probably reliable solution.

**Reply by dwcullop** [reply]
@JakenVeina  I don't think this is a breaking change, I think it's more of a bug fix, because, as you pointed out, if anything downstream wants to perform some action on the object being removed, they're going to be doing it on a Disposed object, and it might be too late at that point.

I think I even hit that bug before, but I just worked around it by moving the `DisposeMany` to be AFTER the `OnItemRemoved` operator, but that might not always be possible.

**Reply by JakenVeina** [reply]
On a technical level, it's an observable (in the colloquial sense) change in the behavior or an API, so it's breaking.

On a practical level, yeah, I struggle to think of a scenario where it matters. It's a little unique on account of being about disposables, because the only way you could "observe" this change that makes disposal occur later is if you were previously using disposables after they'd been disposed. The change is that tomorrow, they might not be disposed yet, when they are today.

Is that sufficient for classifying the change non-breaking? I don't know what's normally the standard to apply, I've never maintained versioning for a large-scale public library.

**Reply by dwcullop** [reply]
Well, [Hyrum's Law](https://www.hyrumslaw.com/) always applies and [every change breaks someone's workflow](https://xkcd.com/1172/), but I honestly think it is acceptable in this case.  There might be places where people have worked around this issue and no longer need to, but I am having a hard time coming up with a situation where something would be broken because the Dispose happens later instead of sooner.  As long as it still happens.

> I that sufficient for classifying the change non-breaking? I don't know what's normally the standard to apply, I've never maintained versioning for a large-scale public library.

It's up to @RolandPheasant and he said he's fine with it.  I think you're all clear to make a PR...

**Reply by RolandPheasant** [reply]
Although this would be a behavioural change, I regard it a bug fix and therefore a minor change, and cannot see that it will break any code.  

Normally I try to apply semantic versioning which is roughly means {major}.{feature}.{minor} where:

major = this is going to break people
feature = new feature
minor = bug fix or optimisation

However as per this discussion there are grey areas but I always balance how likely would problems be caused for consumers.
